jvb/src/main/kotlin/org/jitsi/videobridge/relay/RelayMessageTransport.kt

```diff
@@ -0,0 +1,426 @@
+    override fun sendMessage(dst: Any?, message: BridgeChannelMessage) {
+        super.sendMessage(dst, message) // Log message
+        if (dst is ColibriWebSocket) {
+            sendMessage(dst, message)
+        } else {
+            throw IllegalArgumentException("unknown transport:$dst")
+        }
+    }
+
+    /**
+     * Sends a string via a particular [ColibriWebSocket] instance.
+     * @param dst the [ColibriWebSocket] through which to send the message.
+     * @param message the message to send.
+     */
+    private fun sendMessage(dst: ColibriWebSocket, message: BridgeChannelMessage) {
+        // We'll use the async version of sendString since this may be called
+        // from multiple threads.  It's just fire-and-forget though, so we
+        // don't wait on the result
+        dst.remote?.sendStringByFuture(message.toJson())
+        statisticsSupplier.get().totalColibriWebSocketMessagesSent.incrementAndGet()
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public override fun sendMessage(msg: BridgeChannelMessage) {
+        if (webSocket == null) {
+            logger.debug("No available transport channel, can't send a message")
+            numOutgoingMessagesDropped.incrementAndGet()
+        } else {
+            sentMessagesCounts.computeIfAbsent(
+                msg.javaClass.simpleName
+            ) { k: String? -> AtomicLong() }.incrementAndGet()
+            sendMessage(webSocket, msg)
+        }
+    }
+
+    override val isConnected: Boolean
+        get() = webSocket != null
+
+    val isActive: Boolean
+        get() = outgoingWebsocket != null
+
+    /**
+     * {@inheritDoc}
+     */
+    override fun webSocketConnected(ws: ColibriWebSocket) {
+        synchronized(webSocketSyncRoot) {
+
+            // If we already have a web-socket, discard it and use the new one.
+            if (webSocket != null) {
+                webSocket?.session?.close(200, "replaced")
+            }
+            webSocket = ws
+            sendMessage(ws, createServerHello())
+        }
+        try {
+            notifyTransportChannelConnected()
+        } catch (e: Exception) {
+            logger.warn("Caught an exception in notifyTransportConnected", e)
+        }
+    }
+
+    private fun createServerHello(): ServerHelloMessage {
+        return if (EndpointMessageTransportConfig.config.announceVersion()) {
+            ServerHelloMessage(relay.conference.videobridge.version.toString())
+        } else {
+            ServerHelloMessage()
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    override fun webSocketClosed(ws: ColibriWebSocket, statusCode: Int, reason: String) {
+        synchronized(webSocketSyncRoot) {
+            if (ws == webSocket) {
+                webSocket = null
+                logger.debug { "Web socket closed, statusCode $statusCode ( $reason)." }
+            }
+        }
+        if (outgoingWebsocket != null) {
+            // Try to reconnect.  TODO: how to handle failures?
```

Comment:
Is this going to race with `relay` expiration and re-connect only to be closed as soon as the colibri message arrives?

Reply:
If we haven't called `close` here yet, yeah.  Do you think that's a problem?  Do you think we should listen for the status code and treat certain codes as a deliberate close?

Reply:
I think so, looking at [the docs](https://www.eclipse.org/jetty/javadoc/jetty-9/org/eclipse/jetty/websocket/api/StatusCode.html) it looks like NORMAL and maybe SHUTDOWN can be considered deliberate.